fat/fat2/transactionbatch.go

```diff
@@ -0,0 +1,141 @@
+package fat2
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/Factom-Asset-Tokens/factom"
+	"github.com/Factom-Asset-Tokens/fatd/fat"
+)
+
+// Transaction represents a fat2 transaction, which can be a list of one or more
+// sub-transactions to be executed in order
+type TransactionBatch struct {
+	Version      uint          `json:"version"`
+	Transactions []Transaction `json:"transactions"`
+	fat.Entry
+}
+
+// NewTransactionBatch returns a TransactionBatch initialized with the given entry.
+func NewTransactionBatch(entry factom.Entry) TransactionBatch {
+	return TransactionBatch{Entry: fat.Entry{Entry: entry}}
+}
+
+type transactionBatch TransactionBatch
+
+// MarshalJSON marshals the TransactionBatch content field as JSON, but will raise an error if
+// the batch fails the checks in ValidData()
+func (t TransactionBatch) MarshalJSON() ([]byte, error) {
+	if err := t.ValidData(); err != nil {
+		return nil, err
+	}
+	return json.Marshal(transactionBatch(t))
+}
+
+func (t TransactionBatch) String() string {
+	data, err := t.MarshalJSON()
+	if err != nil {
+		return err.Error()
+	}
+	return string(data)
+}
+
+// UnmarshalEntry unmarshals the Entry content as a TransactionBatch
+func (t *TransactionBatch) UnmarshalEntry() error {
+	return t.Entry.UnmarshalEntry(t)
+}
+
+// MarshalEntry marshals the TransactionBatch into the entry content
+func (t *TransactionBatch) MarshalEntry() error {
+	return t.Entry.MarshalEntry(t)
+}
+
+// Valid performs all validation checks and returns nil if it is a valid batch. This function assumes
+// the struct's entry field is populated.
+func (t *TransactionBatch) Valid() error {
+	err := t.ValidData()
+	if err != nil {
+		return err
+	}
+	if err = t.ValidExtIDs(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ValidData validates all Transaction data included in the batch and returns nil if it is valid.
+// This function assumes that the entry content (or an independent JSON object) has been unmarshaled.
+func (t *TransactionBatch) ValidData() error {
+	// Perform initial validation on all transactions in isolation
+	if t.Version < 1 {
+		return fmt.Errorf("invalid version")
+	}
+	if len(t.Transactions) == 0 {
+		return fmt.Errorf("at least one output required")
+	}
+	for i, tx := range t.Transactions {
+		if err := tx.Valid(); err != nil {
+			return fmt.Errorf("invalid transaction at index %d: %v", i, err)
+		}
+	}
+	return nil
+}
+
+// ValidExtIDs validates the structure of the external IDs of the entry to make
+// sure that it has the correct number of RCD/signature pairs. If no errors are found,
+// it will then validate the content of the RCD/signature pair. ValidExtIDs assumes that
+// the entry content has been unmarshaled and that ValidData returns nil.
+func (t TransactionBatch) ValidExtIDs() error {
```

Comment:
Note that in fat0 and fat1 structural validation of the extids is separated from RCD matching validation.

Reply:
Could you explain the rationale behind that decision? While writing this code, the two felt inherently linked and I couldn't come up with a scenario where I'd want to invoke the functions separately

Reply:
I misspoke here a bit. I said structural validation is separated from RCD matching, which is true. But I also meant that structural and signature validation is separated from RCD matching.

Here is the `Validate()` function for fat0.Transaction: 
https://github.com/Factom-Asset-Tokens/fatd/blob/cdc579c66b3081a9da2dd241cad80abbaaed7c82/fat/fat0/transaction.go#L119-L136

I separated it because validating signatures is generic to all FAT entries, and only depends on a single number from the data: the number of inputs. 

Validating the RCDs depends on whether it is a coinbase transaction or a normal transaction.

I could have put that logic all in the `fat0.Transaction.ValidExtIDs()` function as you did, and then passed through the idKey, but I thought that was more obfuscated than this version of the `fat0.Transaction.Validate()` function above.

Now that I'm revisiting this functions, as per our conversation on discord, I'd like to move ValidData outside of the Unmarshaling and Marshaling functions and put it directly in the `Validate()` functions right below the Unmarshal step. I think this will make each function more specific, i.e. precise, and add to clarity where it counts: the fully validation of the transaction data.

Reply:
Ahh yes now I remember why there was the distinction in fat0 and why I decided to consolidate.

It makes sense for fat0 because the transactions are heterogeneous and it might be a coinbase or a user transaction -- only one of which requiring the ValidRCDs call. However, in fat2, the transactions are homogenous (at least w.r.t. ExtID structure and signing verification) so there isn't a scenario where ExtID structural validation (alongside signature validation) shouldnt happen at the same time as checking that all unique inputs are present in the signature list. Additionally, both actions need the shared result of counting unique input addresses, so I figured it would make sense to combine the two for a slightly simpler API.